apps/system/js/notifications_system_message.js

```diff
@@ -0,0 +1,157 @@
+/* globals BaseModule,
+           Notification
+*/
+
+'use strict';
+
+(function(exports) {
+  var NotificationsSystemMessage = function() {};
+
+  NotificationsSystemMessage.SERVICES = [
+    'handleSystemMessageNotification',
+    'unhandleSystemMessageNotification'
+  ];
+
+  BaseModule.create(NotificationsSystemMessage, {
+    name: 'NotificationsSystemMessage',
+    EVENT_PREFIX: 'systemMessageNotification',
+    DEBUG: false,
+    TRACE: false,
+
+    _handlers: {},
+
+    _start: function nsm_start() {
+      this.registerHandler();
+    },
+
+    registerHandler: function nsm_registerHandler() {
+      this.debug('Installing system app notification message handler');
+      navigator.mozSetMessageHandler('notification',
+        this.processSystemMessage.bind(this));
+    },
+
+    processSystemMessage: function(message) {
+      this.debug('Received system message notification: ' +
+        JSON.stringify(message));
+
+      if (!message.clicked) {
+        return;
+      }
+
+      if (!message.data || !('systemMessageTarget' in message.data)) {
```

Comment:
nit: I don't think we need the `('systemMessageTarget' in message.data)` part of this if, since that is handled below with a more precise error message.